mysql/service_instance.go

```diff
@@ -0,0 +1,434 @@
+		c.client.DebugLogString(fmt.Sprintf("[Debug] : Create Failed %s", serviceInstanceError))
+		return nil, serviceInstanceError
+	}
+
+	return serviceInstance, nil
+}
+
+// GetServiceInstance retrieves the ServiceInstance with the given name.
+func (c *ServiceInstanceClient) GetServiceInstance(getInput *GetServiceInstanceInput) (*ServiceInstance, error) {
+	var serviceInstance ServiceInstance
+	if err := c.getServiceInstanceResource(getInput.Name, &serviceInstance); err != nil {
+		return nil, err
+	}
+
+	return &serviceInstance, nil
+}
+
+// WaitForInstanceReady waits for an instance to be completely initialized and available.
+func (c *ServiceInstanceClient) WaitForServiceInstanceRunning(input *GetServiceInstanceInput, pollingInterval time.Duration, timeoutSeconds time.Duration) (*ServiceInstance, error) {
+	var info *ServiceInstance
+	var getErr error
+
+	err := c.client.WaitFor("service instance to be ready", pollingInterval, timeoutSeconds, func() (bool, error) {
+		info, getErr = c.GetServiceInstance(input)
+		if getErr != nil {
+			return false, getErr
+		}
+		switch s := info.Status; s {
+
+		case ServiceInstanceReady: // Target State
+			return true, nil
+		case ServiceInstanceInitializing:
+			return false, nil
+		case ServiceInstanceStarting:
+			return false, nil
+		default:
+			return false, nil
+		}
+	})
+	return info, err
+}
+
+func (c *ServiceInstanceClient) DeleteServiceInstance(serviceName string) error {
+	if c.Timeout == 0 {
+		c.Timeout = WaitForServiceInstanceDeleteTimeout
+	}
+	if c.PollInterval == 0 {
+		c.PollInterval = WaitForServiceInstanceDeletePollInterval
+	}
+
+	c.client.DebugLogString(fmt.Sprintf("Deleting Instance : %s", serviceName))
+
+	deleteInput := &DeleteServiceInput{}
+
+	deleteErr := c.deleteServiceInstanceResource(serviceName, deleteInput)
+	if deleteErr != nil {
+		c.client.DebugLogString(fmt.Sprintf("[Debug] : Delete Failed %s", deleteErr))
+		return deleteErr
+	}
+
+	// Call wait for instance deleted now, as deleting the instance is an eventually consistent operation
+	getInput := &GetServiceInstanceInput{
+		Name: serviceName,
+	}
+
+	// Wait for instance to be deleted
+	return c.WaitForServiceInstanceDeleted(getInput, c.PollInterval, c.Timeout)
+}
+
+// WaitForServiceInstanceDeleted waits for a service instance to be fully deleted.
+func (c *ServiceInstanceClient) WaitForServiceInstanceDeleted(input *GetServiceInstanceInput, pollingInterval time.Duration, timeoutSeconds time.Duration) error {
+	return c.client.WaitFor("service instance to be deleted", pollingInterval, timeoutSeconds, func() (bool, error) {
+
+		c.client.DebugLogString(fmt.Sprintf("Waiting to destroy instance : %s", input.Name))
+
+		info, err := c.GetServiceInstance(input)
+		if err != nil {
+			if client.WasNotFoundError(err) {
+				// Service Instance could not be found, thus deleted
+				return true, nil
+			}
+			// Some other error occurred trying to get instance, exit
+			return false, err
+		}
+		switch s := info.Status; s {
```

Comment:
Let's add logs under each case so we know what the status is whenever we poll. This is especially useful under the default case when we don't know the specific status is.

Reply:
Done